### PR TITLE
Release W10: Analysis Done

### PR DESCRIPTION
**Pull request by C0DK**
### Description
Finished analysis compleletly and all design documents exist to be able to start development as well as design chapter

### Changelog
 - Analysis Done
 - Syntax done
 - CFG done

## FOR REVIEWER [Do not remove]
A small checklist of things to note when you are doing a review. Make sure these things apply before going into master.
 - Did the PR actually fix the issue
 - Comment anything that doesn't make sense or anything you want clarified. Many comments are always ok
 - Make sure this is ready for exams if this is going into master
 - It's always a good idea having another reviewer look at a PR as well, if it's major. get a buddy.

### Regarding filenames
- all .tex files use pascal case
- all other files [except in /code/] use snake case

### Regarding code [if applicable]
- Check [coding conventions](https://kotlinlang.org/docs/reference/coding-conventions.html)
- Are public methods documented? 
- Are there testcases for new features / methods?
- Have you tried running the source and tested things that cannot be test-cased?

### Regarding Report [if applicable]
- Does each line have a linechange in the sourcecode?
- Does sentences make sense?
- Did you do spellchecking?
- Citations?

### CODE REVIEW

File: code/src/main/kotlin/lexer/Tokens.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    sealed class Type : Token() {
+        class Var : Type()
+        class None : Type()
+        class Text : Type()
+        class Number : Type()
+        class Bool : Type()
+        class Tuple : Type()
+        class List : Type()
+        class Func : Type()
+    }
```
Comment (jones1994) [inline code/src/main/kotlin/lexer/Tokens.kt]: Just a minor question. In the EBNF the "dict" type is included in types, is there a reason why it is not a sublclass in the Type class.

Comment (jones1994) [reply code/src/main/kotlin/lexer/Tokens.kt]: I realize it is a nice to have feature, but the EBNF and the Type class should reflect eachother

Comment (C0DK) [reply code/src/main/kotlin/lexer/Tokens.kt]: i don't think dict needs to be in the EBNF as per our conversations, but i would love to hear everyone elses opinions

File: misc/CFGs/HCLANG_EBNF.txt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+Program --> Commands $
+
+Commands --> {Declaration | Assignment | Expression}
+
+Type --> number
+       | text
+       | tuple
+       | list
+       | dict
```
Comment (C0DK) [inline misc/CFGs/HCLANG_EBNF.txt]: we need to figure out if we are going to use dict or not

File: misc/syntax.txt [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+		void loop() {
+		  digitalWrite(15, HIGH);   // turn the LED on (HIGH is the voltage level)
+		  delay(1000);                       // wait for a second
+		  digitalWrite(15, LOW);    // turn the LED off by making the voltage LOW
+		  delay(1000);                       // wait for a second
+		}
+		###
+
+15)
+	how do we do generic?
```
Comment (C0DK) [inline misc/syntax.txt]: we need to discuss this

File: misc/syntax.txt [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+
+16)
```
Comment (C0DK) [inline misc/syntax.txt]: make sure this is discussed at the next meeting

File: report/2.Analysis/Scope.tex [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+\section{Scope}
+As stated in the above analysis, the purpose of this project is to design, create and implement a programming language for the Arduino platform that is simple for beginners to use, whilst still teaching the fundamentals of programming.
+The language is meant to enable beginners to quickly and easily write
+code that can be executed, without losing the functionality of a high-level language.
+
+It is essential to consider the needed functionality in order to responsibly compose a language for beginners.
+After conducting interviews with gymnasium students, who can be described as beginners at programming, it was decided that the language for this project should be developed as an introductory language for beginners to use, before making a switch to a fully fleshed out programming language.
+
+As such, the language is meant to introduce the thought process used in programming, while ensuring that the user will not have to worry too much about syntax.
+%insert a reference on the next line to the ArduinoInEducation section
```
Comment (C0DK) [inline report/2.Analysis/Scope.tex]: remove comment

File: report/2.Analysis/Scope.tex [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+As was discovered during the visit to the gymnasium class (section \ref{htxinter}), there were a couple of concepts that would be especially important to be wary of when developing the language.
+
+The students seemed to have issues with general syntax, data types, defining functions, control-structures and operator precedence.
+Therefore the developed language will need to have a heightened focus on these particular concepts. 
+
+Adding to these requirements, the project group would also like to reduce the time spent on program-setup as much as possible, allowing the users of the language developed to quickly get from idea to working solution.
+The goal for the language should always be simplicity, comprehensibility and ease of use.
+
+As the group intends to develop an "English-like" language, in the sense that the syntax should look very much like an English sentence, it has been concluded by the group, that almost all operations in the language will be a function.
```
Comment (C0DK) [inline report/2.Analysis/Scope.tex]: this maybe shouldn't be in discussion. not sure - but function decisions is more design than analysis.

Comment (C0DK) [reply report/2.Analysis/Scope.tex]: however the "on top of this" sentence makes sense. but saying that everything should be functions (not almost.. but "all") should probably be in design

File: report/2.Analysis/Scope.tex [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+As the group intends to develop an "English-like" language, in the sense that the syntax should look very much like an English sentence, it has been concluded by the group, that almost all operations in the language will be a function. 
+On top of this, the group found that high-order functions could be both useful and fun to implement and makes sense for an "English-like" language. 
+
+Finally, the group fully intends to develop a full language, with a working compiler, that is production ready, meaning a programming language that will be usable in education.
+
+\section{Problem statement}
+Based on the analysis the following problem statement was defined:
+\begin{center}
+	%1: 
+	%How can a simpler programming language for the Arduino platform, that increases beginners interest and motivation for programming, be developed?
```
Comment (C0DK) [inline report/2.Analysis/Scope.tex]: what is this? why the comments?

File: report/2.Analysis/Scope.tex [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	%2:
+	How can a programming language that eases development for the Arduino platform be developed?
+
+\begin{itemize}
+	\item How can simplicity and a quick development process be ensured? %means that we integrate our processes and whatnot in the project
+	\item How can the basic programming concepts be taught, in a manner that ensures, that the user can recognize and use the concepts in more advanced languages?
+	\item How can programming language be implemented in a manner that uses functions for most operations? 
+	%Make a subitem regarding ensuring functions are an integral part of the language
```
Comment (C0DK) [inline report/2.Analysis/Scope.tex]: so this is not done? if "make a sub item" is in this?

File: misc/syntax.txt (hunk 3 above)
Comment (jones1994) [reply misc/syntax.txt]: Has been added to next agenda

File: misc/syntax.txt (hunk 4 above)
Comment (jones1994) [reply misc/syntax.txt]: Has been added to next agenda

File: report/2.Analysis/ProblemStatement.tex [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+\begin{itemize}
+	\item How can simplicity and a quick development process be ensured? %means that we integrate our processes and whatnot in the project
+	\item How can the basic programming concepts be taught, in a manner that ensures, that the user can recognize and use the concepts in more advanced languages?
+	\item How can programming language be implemented in a manner that uses functions for most operations? 
+	%Make a subitem regarding ensuring functions are an integral part of the language
+\end{itemize}
+\end{center}
+%Dont know if we should add more here.
+%If you have any ideas for different problem statements, do feel free to add them :)
```
Comment (jones1994) [inline report/2.Analysis/ProblemStatement.tex]: All comments should be removed

File: report/2.Analysis/Scope.tex (hunk 6 above)
Comment (jones1994) [reply report/2.Analysis/Scope.tex]: I think it is fine here. But I see your point. Further argument for this decision should be in the design documentation.

File: code/src/main/kotlin/lexer/Tokens.kt (hunk 1 above)
Comment (Lambcreek) [reply code/src/main/kotlin/lexer/Tokens.kt]: I think it should be removed from the EBNF. It doesn't seem important to our goal, and I don't think we will learn much from including it, so it's just more work for no reason.

File: report/2.Analysis/Scope.tex (hunk 6 above)
Comment (nillas12) [reply report/2.Analysis/Scope.tex]: That is being done in language specification, seen as solved.

**Pull request merged**